apiserver/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,141 @@
+---
+title: OAuth-Proxy-Proxy-Support
+authors:
+  - "@deads2k"
+reviewers:
+  - "@sur"
+  - "@bparees"
+  - "@enj"
+  - "@sttts"
+approvers:
+  - "@sttts"
+creation-date: 2019-09-18
+last-updated: 2019-09-18
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# OAuth Proxy Proxy Support
+
+We should create zero-config method of enabling proxy-support for the oauth-proxy to enable non-expert teams to be secure
+when proxies are configured in a cluster.
+The `oauth-proxy` is used to front many webservers to provide authentication and authorization protection without directly
+building it into their binaries. This is common, easy, and used by people without a deep understand of how OpenShift is
+wired together.  They are not likely to know how to reactively wire proxy configuration and ca-bundles into their 
+side-cars.  Because it is re-used by directly embedding the `oauth-proxy` into a `PodSpec`, any changes to the way it is
+invoked will cause adoption difficulties.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+In order to function, the `oauth-proxy`
+ 1. Gets the oauth-server location using the kube-apiserver's internal IP: `https://172.30.0.1/.well-known/oauth-authorization-server`
+ 2. Uses the result to find the token endpoint.
+ 3. Hits the *external* token endpoint `https://oauth-openshift.apps.<cluster-dns-prefix>.openshift.com/oauth/token`.
+
+This flow requires hitting a potentially external router, which requires honoring proxy settings.
+
+## Motivation
+
+See summary in first section.
+
+### Goals
+
+1. Zero-config proxy support that dynamically updates as proxy settings change.
+2. Support for all proxy modes, including MITM.
+
+### Non-Goals
+
+1. Making generic code to rewrite proxy transport configuration.
+
+## Proposal
+
+The `oauth-proxy` can be written to use targeted list/watch against the proxy.config.openshift.io and 
+`oc get -n openshift-config-managed configmaps/proxy-ca` to build a proxy transport.  Using standard informers with single item
+watches allows RBAC to only expose GET for individual items.  It also means that an efficient swap mechanism can be built
+to ensure performant behavior.
+
+This is an intricate thing to write, so we avoided it for every other proxy component, but this one is a component that is
+re-used by numerous external teams by directly embedding it into a pod spec.  This changes the tradeoff, so it becomes worthwhile
+to write this code.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. If HTTP env vars are set, the dynamic proxy-support code must be disabled.  The user is in charge of what the `oauth-proxy`
+ does in these cases.
```

Comment:
MITM proxy CA is only the first use case we have for the trusted-ca-bundle, i think we envision it expanding at some point to "This is the set of CAs you should trust".  If the oauth-proxy has to talk to external auth providers, i could envision it needing additional CAs to trust at that point and i'd expect trusted-ca-bundle might be the mechanism we use to provide those CAs.

Even if that isn't the mechanism, there will be some similar mechanism from which the oauth-proxy will need to read/consume additional CAs for that use case.  So i don't think we should restrict this behavior/capability to "only when proxy env vars are set".

Reply:
> Even if that isn't the mechanism, there will be some similar mechanism from which the oauth-proxy will need to read/consume additional CAs for that use case. So i don't think we should restrict this behavior/capability to "only when proxy env vars are set".

I think you've confused the special cases (disable in case of...) with the general "how do you do this" from the proposal above.

To disable the feature, you have to work with what you can observe from the binary.  "someone injected different trust bundles" can't be observed.

Reply:
no, i just misread "disabled" as "enabled" for some reason :(

this makes sense.